### ISSUE TITLE
When caret is in a list element, shift-enter should move to new line but not add bullet

### ISSUE THREAD

**Issue by vjpr** [vsch/idea-multimarkdown#516]
**Expected**

```
- foo<shift-enter>
  |
- bar
```

**Actual**

```
- foo<shift-enter>
- |
- bar
```

**Comment by vsch** [comment]
@vjpr, see #192

**Comment by vjpr** [comment]
I think my issue is different.

I want `shift+enter` to create sub text without a bullet, but not stop bulleting.

**Comment by vsch** [comment]
@vjpr, shift enter is not a separate handler in the IDE but a combination of cursor move and enter combination so it is not possible to have one insert a list item and the other not.

However, as a workaround hitting enter on an empty list item removes the item prefix. This is the best I could come up with considering the limitations.

**Comment by vjpr** [comment]
Okay so the workaround is okay:

```
- bar
  - foo<enter>
  - <enter>
  <tab>  |
```

**Comment by vjpr** [comment]
But there seems to be `Editor actions > Start new line` set to `Shift+Enter` in the Keymap. If Markdown exposed a Shift+Enter action then we could bind to this no?

**Comment by vsch** [comment]
The Plugin does not override the action but implements an EnterHandler, which is the way the IDE expects customizations of the ENTER to be implemented.

**Comment by vsch** [comment]
@vjpr, it sounds like you are doing manual text formatting. Why are you not using the wrap on typing and element or document format to have the text reflow within a list item?

**Comment by vjpr** [comment]
I want to be able to write documents with sub-lines like so:

- test
   _this is a sub-line_
- test
- test

**Comment by vsch** [comment]
@vjpr, are you using the hard line break (two or more spaces at the end of the line) to separate the sub-line from the main item?

If you are not then this is only in the markdown source but not in the rendered HTML.

If you are using the hard-break spaces then that could be added to the logic for handling ENTER. If the line you are on end is two or more spaces then ENTER will not add the item marker.

**Comment by vjpr** [comment]
> are you using the hard line break (two or more spaces at the end of the line) to separate the sub-line from the main item?

I wasn't. And did not know about [hard line breaks](http://markdown-guide.readthedocs.io/en/latest/basics.html#line-return).

When the profile is set to `Github Markdown Documents` two trailing spaces are required.

When the profile is set to `Github Comments`, the document renders how I like it.

> If you are using the hard-break spaces then that could be added to the logic for handling ENTER.

Yep, this is what would work best for me. Shift+Enter to create a line return without a bullet.

> If the line you are on end is two or more spaces then ENTER will not add the item marker.

Are you saying this is how it works presently (it doesn't), or how it *should* work?

**Comment by vsch** [comment]
@vjpr, GitHub comments has HARD BREAKS enabled, which treats all line breaks as hard breaks.

Usually, the parser configuration is set for how the document will be rendered outside the IDE, for example if it is on GitHub then it will be rendered as GitHub Markdown Document not as a comment and implicit hard breaks will not be rendered.

If comment mode works for your use case then it is not an issue.

**Comment by vjpr** [comment]
This issue is just about the `shift+enter` behavior though.

If hard breaks enabled, add two spaces, new line, **indent to left-most text**.

If hard breaks not enabled, new line, **indent to left-most text**.

**Comment by vsch** [comment]
@vjpr, since there is nothing I can do about shift+enter without affecting enter then an intelligent workaround is the only option.

What I am proposing is that if the current line ends is two or more spaces then enter will indent the caret on the next line (if following a list or other indenting item), whether HARD-WRAPS option is enabled in the parser or not.  

The two spaces continuing  a list item text makes sense based on markdown syntax. However, using hard-wraps parser mode to disable this only applies to your use case and making it behave this way for everyone would not be welcome by other users.

Once this behavior for two spaces for ENTER is added then you can define you own macro that will move to the end of the line, add two spaces and then execute ENTER. Assigning it to SHIFT+ENTER will give you the functionality that you want.

**Comment by vjpr** [comment]
> since there is nothing I can do about shift+enter

In the keymap I see this:

![screen shot 2017-11-04 at 12 46 18 pm](https://user-images.githubusercontent.com/281413/32405073-73756c74-c15e-11e7-97d1-5018cd4607b8.png)

So I was wondering why you cannot create an action that can be "new line and indent". Surely key-mappable actions can check the position of the cursor and do multiple things...or is this where I am wrong?

> that will move to the end of the line

I just realized that `shift+enter` will create a new line regardless of where the cursor is. Maybe this is the confusion. Maybe it would be better as `alt+enter` or `cmd+enter` then?

Basically, if you have Microsoft Word, I want the same list experience which that offers.

**Comment by vjpr** [comment]
Like this: ![kapture 2017-11-04 at 13 29 28](https://user-images.githubusercontent.com/281413/32405369-5a942942-c164-11e7-9a5e-c8d3d50d0107.gif)

**Comment by vsch** [comment]
> So I was wondering why you cannot create an action that can be "new line and indent". Surely key-mappable actions can check the position of the cursor and do multiple things...or is this where I am wrong?

The answer is simple. I only have so much available time and have to prioritize requests. I do this based on usefulness to many users. As much as I would like to have every user's wishes satisfied, it is not realistic from my time management perspective.

Once the two spaces at end of line context sensitivity is implemented for ENTER, you will be able to create a macro that will do what you want without needing the plugin to create a new action.

With the feature request to have TAB and SHIFT+TAB perform context sensitive functions, including indent/un-indent list items, I think most of Word type functionality for list manipulation will be available in the plugin.

For now there are Increase/Decrease list level actions that perform this as separate actions. I agree it is less convenient than TAB and SHIFT+TAB but overriding actions assigned to standard keys has unforeseen consequences on behavior that needs time to debug and fix. That is the reason I did not jump on implementing this feature.

**Comment by vjpr** [comment]
> I only have so much available time and have to prioritize requests.

If this is the case, then this feature does not matter so much to me. 

But would it be possible to please implement https://github.com/vsch/idea-multimarkdown/issues/469 quickly? This is a daily frustration much more than anything else I have requested. If you can do this, then I am happy for a long time. I would 100% implement myself if it were possible...but as you have mentioned it is not right now.
